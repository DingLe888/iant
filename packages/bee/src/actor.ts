export default class Actor {
  defaultState(): Object {
    return {};
  }
}
