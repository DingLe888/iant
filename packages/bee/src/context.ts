import React from 'react';
import Store from './store';

/**
 * 创建store的上下文
 */
export const StoreContext = React.createContext({} as Store);
