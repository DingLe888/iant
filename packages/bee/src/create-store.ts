import { IStoreProps } from './types';

export default function createStore(props: IStoreProps) {
  return props;
}
