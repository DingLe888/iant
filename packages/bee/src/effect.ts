import produce from 'immer';
import { IReducer, IReducerProps } from './types';

export default function effect<T>(obj: IReducerProps<T>) {
  const immerHandler = {} as IReducer<T>;
  //all handler wrapper produce method
  for (let key in obj) {
    const handler = obj[key];
    immerHandler[key] = produce<T>(handler as any);
  }
  return immerHandler;
}
