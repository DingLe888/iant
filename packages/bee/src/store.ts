import produce from 'immer';
import { QueryLang } from './ql';
import { IReducer, IStoreProps, TPath, TSubscriber } from './types';
import { getPathVal, isArray, isStr } from './util';

export default class Store<T = Object> {
  constructor(props: IStoreProps<T>) {
    const { state = {}, ql, reducer } = props;
    this._state = state;
    this._cache = {};
    this._subscribe = [];
    this._ql = ql || this.ql();
    this._reducer = reducer;

    //merge rx
    const rx = this._computeQL();
    this._state = {
      ...this._state,
      ...rx
    };
  }

  private _state: Object;
  private _subscribe: Array<TSubscriber>;
  private _ql: { [name: string]: QueryLang };
  private _cache: { [key: number]: Array<any> };
  private _reducer: IReducer<T>;

  ql() {
    return {};
  }

  _computeQL() {
    return Object.keys(this._ql).reduce((r, k) => {
      const ql = this._ql[k];
      r[k] = this.bigQuery(ql);
      return r;
    }, {});
  }

  dispatch(action: string, params?: Object) {
    if (!this._reducer) {
      return;
    }
    const handler = this._reducer[action];
    const state = handler(this._state as any, params);
    if (state != this._state) {
      this._state = state;
      const rx = this._computeQL();
      this._state = {
        ...this._state,
        ...rx
      };
      for (let subscribe of this._subscribe) {
        subscribe(this._state);
      }
    }
  }

  getState() {
    return Object.freeze(this._state);
  }

  setState(callback: (data: T) => void) {
    const state = produce(this._state, callback as any);
    if (state != this._state) {
      this._state = state;
      const rx = this._computeQL();
      this._state = {
        ...this._state,
        ...rx
      };
      for (let subscribe of this._subscribe) {
        subscribe(this._state);
      }
    }
  }

  bigQuery(query: TPath | QueryLang) {
    if (isStr(query) || isArray(query)) {
      return getPathVal(this._state, query);
    } else if (query instanceof QueryLang) {
      let isChanged = false;
      const { id, deps, handler } = query.meta();
      //init cache
      this._cache[id] || (this._cache[id] = []);
      const len = deps.length;

      //计算pathVal
      deps.forEach((dep, i) => {
        const val = this.bigQuery(dep);
        if (val !== this._cache[id][i]) {
          isChanged = true;
        }
        this._cache[id][i] = val;
      });

      if (isChanged) {
        const depVal = this._cache[id].slice(0, len);
        const result = handler(...depVal);
        this._cache[id][len] = result;
        return result;
      } else {
        return this._cache[id][len];
      }
    }
  }

  subscribe(callback: TSubscriber) {
    let index = this._subscribe.indexOf(callback);

    if (index === -1) {
      this._subscribe.push(callback);
      index = this._subscribe.indexOf(callback);
    }

    return () => {
      this._subscribe.splice(index, 1);
    };
  }
}
