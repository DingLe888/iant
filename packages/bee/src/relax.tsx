import React from 'react';
import { StoreContext } from './context';
import Store from './store';
import { isArray, isObj, isStr } from './util';

export interface IRenderProps {
  relaxProps: any;
  [name: string]: any;
}

export interface IProps {
  relaxProps: Array<any>;
  render: (props: IRenderProps) => React.ReactElement<Object>;
  [name: string]: any;
}

export default class Relax extends React.Component<IProps> {
  static displayName = 'RelaxContext';
  static contextType = StoreContext;

  constructor(props: IProps, ctx: Store) {
    super(props);

    this._context = ctx;
    this._isMounted = false;
    this._relaxPropsMapper = this._reduceRelaxPropsMapper();
    this._relaxProps = this._computeRelaxProps();

    if (!this._isAllFn()) {
      this._unsubsciber = this._context.subscribe(this._handleSubscribe);
    }
  }

  componentDidMount() {
    this._isMounted = true;
  }

  shouldComponentUpdate(nextProps: Object) {
    const nextRelaxProps = this._computeRelaxProps();
    if (nextProps !== this.props || nextRelaxProps != this._relaxProps) {
      this._relaxProps = nextRelaxProps;
      return true;
    } else {
      return false;
    }
  }

  componentWillUnmount() {
    this._isMounted = false;
    this._unsubsciber && this._unsubsciber();
  }

  componentWillUpdate() {
    this._isMounted = false;
  }

  componentDidUpdate() {
    this._isMounted = true;
  }

  render() {
    const { render, relaxProps, ...other } = this.props;
    return render({
      ...other,
      relaxProps: {
        ...this._relaxProps
      }
    });
  }

  private _isMounted: boolean;
  private _context: Store;
  private _relaxPropsMapper: Object;
  private _relaxProps: Object;
  private _unsubsciber: Function;

  _computeRelaxProps() {
    const store: Store = this._context;
    const mapper = this._relaxPropsMapper;
    const relaxData = {};

    for (let prop in mapper) {
      const val = mapper[prop];
      //如果是数组，取最后一个名字，作为可以
      if (isArray(val)) {
        relaxData[prop] = store.bigQuery(val);
      }
      //如果是字符串
      else if (isStr(prop)) {
        if (prop === 'dispatch' || prop === 'setState') {
          relaxData[prop] = store[prop];
        } else {
          relaxData[prop] = store.bigQuery(val);
        }
      }
    }

    return relaxData;
  }

  _reduceRelaxPropsMapper() {
    const { relaxProps } = this.props;
    const relaxData = Object.create(null);

    for (let prop of relaxProps) {
      if (isArray(prop)) {
        const len = prop.length;
        const last = prop[len - 1];
        relaxData[last] = prop;
      } else if (isStr(prop)) {
        if (prop.indexOf('.') != -1) {
          const arr = prop.split('.');
          const len = arr.length;
          const last = arr[len - 1];
          relaxData[last] = arr;
        } else {
          relaxData[prop] = prop;
        }
      } else if (isObj(prop)) {
        for (let key in prop) {
          if (prop.hasOwnProperty()) {
            const val = prop[key];
            relaxData[key] = val;
          }
        }
      }
    }

    return relaxData;
  }

  _handleSubscribe = state => {
    if (this._isMounted) {
      this.setState(() => state);
    }
  };

  _isAllFn() {
    const mapper = this._relaxPropsMapper;
    //如果所有注入的属性都是setState或者dispatch，就不去监听subscribe
    for (let key in mapper) {
      const val = mapper[key];
      if (isStr(val) && (val === 'setState' || val === 'dispatch')) {
        continue;
      } else {
        return false;
      }
    }
    return true;
  }
}
