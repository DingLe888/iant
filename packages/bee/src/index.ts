import Provider from './provider';
import reducer from './reducer';
import Relax from './relax';
import { createStore } from './store';

export { createStore, Relax, reducer, Provider };
