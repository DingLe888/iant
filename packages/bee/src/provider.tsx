import React from 'react';
import { StoreContext } from './context';
import { Store } from './store';
import { IProviderProps } from './types';

export default class Provider extends React.Component<IProviderProps> {
  constructor(props: IProviderProps) {
    super(props);
    this._store = this.props.store();
  }

  private _store: Store;

  componentWillMount() {
    this.props.onWillMount && this.props.onWillMount();
  }

  componentDidMount() {
    this.props.onMounted && this.props.onMounted(this._store);
  }

  componentDidUpdate() {
    this.props.onUpdated && this.props.onUpdated();
  }

  render() {
    return (
      <StoreContext.Provider value={this._store}>
        {this.props.children}
      </StoreContext.Provider>
    );
  }
}
