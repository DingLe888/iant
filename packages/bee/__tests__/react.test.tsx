import React from 'react';
import render from 'react-test-renderer';
import createStore from '../src/create-store';
import Provider from '../src/provider';
import Relax from '../src/relax';
import { isFn } from '../src/util';

interface IRelaxProps {
  id: number;
  lname: string;
  name: string;
  dispatch: Function;
  setState: Function;
}

const store = createStore({
  state: {
    name: 'test',
    list: [{ id: 1, name: 'test' }]
  }
});

const TestRelax = () => (
  <Relax
    relaxProps={[
      //
      ['list', 0, 'id'],
      { lname: 'list.0.name' },
      'name',
      'dispatch',
      'setState'
    ]}
    render={({ relaxProps }: { relaxProps: IRelaxProps }) => (
      <div>
        {relaxProps.id}
        {relaxProps.name}
        {relaxProps.lname}
        {isFn(relaxProps.dispatch) && 'yes'}
        {isFn(relaxProps.setState) && 'yes'}
      </div>
    )}
  />
);

const TestApp = () => (
  <Provider store={store as any}>
    <TestRelax />
  </Provider>
);

it('test init', () => {
  const tree = render.create(<TestApp />);
  expect(tree).toMatchSnapshot();
});

it('test no subscribe', () => {
  // let _ref = React.createRef();
  let _ref = null as any;

  const RelaxApp = () => (
    <Relax
      ref={ref => {
        _ref = ref;
      }}
      relaxProps={['dispatch', { setState: 'setState' }]}
      render={() => <div />}
    />
  );
  const TestApp = () => (
    <Provider store={store as any}>
      <RelaxApp />
    </Provider>
  );

  const tree = render.create(<TestApp />);
  expect(tree).toMatchSnapshot();
  expect(_ref._isAllFn()).toEqual(true);
});

it('test change state', () => {
  const RelaxApp = () => (
    <Relax
      relaxProps={['list.0.id']}
      render={({ relaxProps }) => <div>{relaxProps.id}</div>}
    />
  );

  let _ref = null as any;
  const TestApp = () => (
    <Provider
      store={store as any}
      ref={ref => {
        _ref = ref;
      }}
    >
      <RelaxApp />
    </Provider>
  );

  const tree = render.create(<TestApp />);
  expect(tree.toJSON()).toMatchSnapshot();

  _ref._store.setState(state => {
    state.list[0].id = 2;
  });

  expect(tree.toJSON()).toMatchSnapshot();
});
