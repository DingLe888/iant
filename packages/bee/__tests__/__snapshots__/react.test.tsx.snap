// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test change state 1`] = `
<div>
  1
</div>
`;

exports[`test change state 2`] = `
<div>
  2
</div>
`;

exports[`test init 1`] = `
<div>
  1
  test
  yes
  yes
</div>
`;

exports[`test no subscribe 1`] = `<div />`;
